### PR TITLE
support llava1.5 lora finetuning.

### PR DESCRIPTION
**Pull request by lkk12014402**
# What does this PR do?

add llava1.5 finetuning and add an example

### CODE REVIEW

**Comment by lkk12014402** [comment]
### finetuning on gaudi

#### Before optimization 
image

#### After optimization (this pr)

image

### finetuning on a100

image

**Comment by yao-matrix** [comment]
@libinta , pls help review, validated pass in 1.19.0-410 build, thx.

**Comment by yafshar** [comment]
@lkk12014402 thanks for your contribution. 

Can you explain why you created a new script, `run_llava_lora_finetune.py`, which is very similar to `run_image2text_lora_finetune.py`, instead of modifying the original script? Introducing new code can increase the possibility of errors and make maintenance more challenging.

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 1]
```diff
@@ -136,50 +137,48 @@ def forward(
...
-                image_outputs = self.vision_tower(
-                    pixel_values,
-                    output_hidden_states=True,
-                    use_flash_attention=use_flash_attention,
-                    flash_attention_recompute=flash_attention_recompute,
-                )
-                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
-                selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
+        # 1. Extra the input embeddings
+        inputs_embeds = self.get_input_embeddings()(input_ids)
```
Comment (yafshar) [inline optimum/habana/transformers/models/llava/modeling_llava.py:155]: @lkk12014402 this fails if you have `inputs_embeds` as input and not `input_ids`. I understand that you used the previous code, but the logic is not correct!

Comment (yafshar) [inline optimum/habana/transformers/models/llava/modeling_llava.py:155]:
```suggestion
        if (input_ids is None) ^ (inputs_embeds is not None):
            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")

        if pixel_values is not None and inputs_embeds is not None:
            raise ValueError(
                "You cannot specify both pixel_values and inputs_embeds at the same time, and must specify either one"
            )

        if inputs_embeds is None:
            inputs_embeds = self.get_input_embeddings()(input_ids)
```

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 2]
```diff
@@ -220,20 +219,52 @@ def forward(
...
                 inputs_embeds=inputs_embeds,
-                vision_feature_layer=vision_feature_layer,
-                vision_feature_select_strategy=vision_feature_select_strategy,
-                labels=labels,
                 use_cache=use_cache,
                 output_attentions=output_attentions,
                 output_hidden_states=output_hidden_states,
                 return_dict=return_dict,
+                cache_position=cache_position,
+                # TODO: from Transformers v4.45, `generate` sets `num_logits_to_keep` to 1 if not given, which we don't want here
```
Comment (yafshar) [inline optimum/habana/transformers/models/llava/modeling_llava.py:232]: @lkk12014402 we are on Transformers v4.45.2, I think we should fix the TODO and remove the comment

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 3]
```diff
@@ -220,20 +219,52 @@ def forward(
...
+                # Flatten the tokens
+                loss_fct = nn.CrossEntropyLoss()
+                loss = loss_fct(
+                    shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1).to(shift_logits.device)
+                )
+
+            if not return_dict:
+                output = (logits,) + outputs[1:]
+                return (loss,) + output if loss is not None else output
+            # print(loss)
```
Comment (yafshar) [inline optimum/habana/transformers/models/llava/modeling_llava.py:259]:
```suggestion
```

**Comment by yafshar** [comment]
@lkk12014402 in your performance comparison above, what kind of optimization you did? I do not see any optimization. Can you elaborate?

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 1 above)
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: fixed

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 2 above)
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: updated

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 3 above)
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: deleted

**Comment by lkk12014402** [comment]
> @lkk12014402 thanks for your contribution.
> 
> Can you explain why you created a new script, `run_llava_lora_finetune.py`, which is very similar to `run_image2text_lora_finetune.py`, instead of modifying the original script? Introducing new code can increase the possibility of errors and make maintenance more challenging.

hi, llava's Processor, DataCollator, evaluation code, padding style are different from the `run_image2text_lora_finetune.py`. If we merge the llava's code into the file `run_image2text_lora_finetune.py`, some `if-else` code will be ingested into the file.  And I think each model maybe have its own Processor and DataCollator, or suitable dataset,  So it is clear to create a new script.

**Comment by lkk12014402** [comment]
> @lkk12014402 in your performance comparison above, what kind of optimization you did? I do not see any optimization. Can you elaborate?

padding inputs, [datacollator](https://github.com/huggingface/optimum-habana/pull/1487/files#diff-c9ba9fe6cc955c1616632748d248b039e0925667965b5586a14a4d06e7ba39a0R264)

**Comment by yafshar** [comment]
@lkk12014402 can you also add the tests results? Does your change breaks any unit tests (test_image_to_text_example)?

**Comment by yafshar** [comment]
I ran the test

main

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_image_to_text_example.py -v -s -k llava

10 passed, 4 deselected in 4419.67s (1:13:39)
```

this PR
```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_image_to_text_example.py -v -s -k llava

10 passed, 4 deselected in 2320.45s (0:38:40)
```

**Review by yafshar — APPROVED** [review]
I am still not sure about adding a new script but other than that it LGTM!

@regisss would you please check this PR.

**Comment by lkk12014402** [comment]
> I am still not sure about adding a new script but other than that it LGTM!
> 
> @regisss would you please check this PR.

hi, @yafshar  I have merged the 2 scripts.

Thanks~

**Comment by yafshar** [comment]
@lkk12014402 thanks. Can you do the full test to make sure your changes did not break anything?

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_image_to_text_example.py
```

**Comment by yafshar** [comment]
OK, I ran the tests. They seem to be fine. Would you please try some other examples

main branch
```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_image_to_text_example.py -v -s 
14 passed in 1957.63s (0:32:37)
```

this PR
```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_image_to_text_example.py -v -s 
14 passed in 1428.22s (0:23:48)
```

**Comment by lkk12014402** [comment]
hi, @regisss  please review~

**Review by vidyasiv — CHANGES_REQUESTED** [review]
based on regiss feedback: please add/update relevant test(s) for new script/model: https://github.com/huggingface/optimum-habana/blob/main/tests/test_image_to_text_example.py

**Comment by libinta** [comment]
@lkk12014402 do you still want to this change? If so, please revise based on comments

**Review by vidyasiv — APPROVED** [review]

**Comment by emascarenhas** [comment]
@lkk12014402 , Please revise/update for any code changes.

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 4]
```diff
@@ -220,20 +237,50 @@ def forward(
...
+                    shift_logits = logits[..., :-1, :][shift_attention_mask.to(logits.device) != 0].contiguous()
+                    shift_labels = labels[..., 1:][shift_attention_mask.to(labels.device) != 0].contiguous()
+                else:
+                    shift_logits = logits[..., :-1, :].contiguous()
+                    shift_labels = labels[..., 1:].contiguous()
+                # Flatten the tokens
+                loss_fct = nn.CrossEntropyLoss()
+                loss = loss_fct(
+                    shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1).to(shift_logits.device)
+                )
```
Comment (vidyasiv) [inline optimum/habana/transformers/models/llava/modeling_llava.py:271]: If this PR goes in after transformers_future, we need to refactor and use loss_function() and recheck performance and tests

Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py:271]: hi, @vidyasiv  the code specific the transformers version 4.45.2 https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/llava/modeling_llava.py#L536

File: examples/image-to-text/README.md [hunk 5]
```diff
@@ -375,6 +375,40 @@ python3 ../gaudi_spawn.py \
     --lora_target_modules '".*(language_model).*(down_proj|gate_proj|up_proj|k_proj|q_proj|v_proj|o_proj).*$"'
 ```
 
+Here are single card training command examples for llava-hf/llava-1.5-7b-hf.
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply examples/image-to-text/README.md]: updated

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 4 above)
Comment (vidyasiv) [reply optimum/habana/transformers/models/llava/modeling_llava.py:271]: @lkk12014402 ,  For now no change needed, lets assume we merge before this pr: https://github.com/huggingface/optimum-habana/pull/1698

File: examples/image-to-text/README.md [hunk 6]
```diff
@@ -168,6 +168,94 @@ python3 ../gaudi_spawn.py \
     --lora_target_modules '".*(language_model).*(down_proj|gate_proj|up_proj|k_proj|q_proj|v_proj|o_proj).*$"'
 ```
 
+
+The single card training command for llava-hf/llava-1.5-7b-hf is similar.
```
Comment (vidyasiv) [inline examples/image-to-text/README.md:172]: The idea behind adding this line was to also remove lines 174-204, i.e we dont need the command written down again.

Comment (emascarenhas) [reply examples/image-to-text/README.md:172]: @lkk12014402 , Please remove the command and only include the comment. If there are some options that are different from llama-3.x you can call out those options in the text. Thanks.

Comment (vidyasiv) [reply examples/image-to-text/README.md:172]: @lkk12014402 please update

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 4 above)
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py:271]: ok, thanks~

File: examples/image-to-text/README.md (hunk 6 above)
Comment (lkk12014402) [reply examples/image-to-text/README.md:172]: updated

File: examples/image-to-text/README.md [hunk 7]
```diff
@@ -168,6 +168,63 @@ python3 ../gaudi_spawn.py \
...
+```bash
+QUANT_CONFIG=./quantization_config/maxabs_quant_scale_format_const.json python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_pipeline.py \
+    --model_name_or_path llava-hf/llava-v1.6-mistral-7b-hf \
+    --image_path "https://llava-vl.github.io/static/images/view.jpg" \
+    --use_hpu_graphs \
+    --bf16 \
+    --use_flash_attention \
+    --flash_attention_recompute
+```
+
```
Comment (libinta) [inline examples/image-to-text/README.md:226]: no need for new cmd as we already have it

Comment (lkk12014402) [reply examples/image-to-text/README.md]: removed

**Review by vidyasiv — APPROVED** [review]
LGTM

File: examples/image-to-text/README.md (hunk 7 above)
Comment (emascarenhas) [reply examples/image-to-text/README.md]: @lkk12014402 , Please remove the extra new command. Instead just write a text statement indicating that the model is supported and any special options needed to run it.

**Comment by emascarenhas** [comment]
@lkk12014402 , Can you also redo full test (see command above from Dec 6) with the 1.20 release?

**Review by regisss — COMMENTED** [review]
Let's rather use `config.model_type` everywhere, see my comments below.
Also, can you add a test for fine-tuning Llava with LoRA please?

**Review by libinta — COMMENTED** [review]
@lkk12014402 can you add a UT?

**Review by regisss — APPROVED** [review]
LGTM!

I also added a regression test.
